File: app/src/main/java/app/kamy/saatApp/domain/model/DoaModels.kt
package app.kamy.saatApp.domain.model

import kotlinx.serialization.KSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.JsonDecoder
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonPrimitive

@OptIn(kotlinx.serialization.ExperimentalSerializationApi::class)
object FlexibleTranslationSerializer : KSerializer<String?> {
    override val descriptor: SerialDescriptor =
        PrimitiveSerialDescriptor("FlexibleTranslationSerializer", PrimitiveKind.STRING)

    override fun serialize(encoder: Encoder, value: String?) {
        if (value != null) {
            encoder.encodeString(value)
        } else {
            encoder.encodeNull()
        }
    }

    @OptIn(kotlinx.serialization.ExperimentalSerializationApi::class)
    override fun deserialize(decoder: Decoder): String? {
        val input = decoder as? JsonDecoder ?: return decoder.decodeString()
        return when (val element = input.decodeJsonElement()) {
            is JsonNull -> null
            is JsonPrimitive -> element.contentOrNull
            is JsonObject -> {
                element["id"]?.jsonPrimitive?.contentOrNull
                    ?: element["ms"]?.jsonPrimitive?.contentOrNull
                    ?: element["en"]?.jsonPrimitive?.contentOrNull
            }
            else -> null
        }
    }
}

@Serializable
data class DoaCategory(
    val id: String? = null,
    val name: String? = null,
    val slug: String? = null,
    val photo: String? = null
)

@Serializable
data class DoaItem(
    val id: String? = null,
    val title: String? = null,
    val arabic: String? = null,
    val latin: String? = null,
    @Serializable(with = FlexibleTranslationSerializer::class)
    val translation: String? = null,
    val notes: String? = null,
    val fawaid: String? = null,
    val source: String? = null,
    val category: String? = null,
    val categories: DoaCategory? = null
)

@Serializable
data class DoaListResponse(
    val success: Boolean? = null,
    val message: String? = null,
    val data: List<DoaItem>? = null
)

@Serializable
data class DhikrContentItem(
    val id: String? = null,
    val arabic: String? = null,
    val latin: String? = null,
    @Serializable(with = FlexibleTranslationSerializer::class)
    val translation: String? = null,
    val fawaid: String? = null,
    val notes: String? = null,
    val source: String? = null,
    @SerialName("repeat_count") val repeatCount: Int? = null
)

@Serializable
data class DhikrBundle(
    val title: String? = null,
    val category: String? = null,
    @SerialName("compiled_by") val compiledBy: String? = null,
    val content: List<DhikrContentItem>? = null
)

@Serializable
data class DhikrListResponse(
    val success: Boolean? = null,
    val message: String? = null,
    val data: List<DhikrBundle>? = null
)

@Serializable
data class DoaCategoriesResponse(
    val success: Boolean? = null,
    val message: String? = null,
    val data: List<DoaCategory>? = null
)

enum class DoaCatalogKind { DOA, DHIKR }

data class DoaCatalogEntry(
    val slug: String,
    val title: String,
    val kind: DoaCatalogKind
)

File: iosApp/Saat/Domain/Models/DoaModels.swift
//
//  DoaModels.swift
//  Saat
//
//  Created by Elmee on 25/06/2026.
//  Copyright © 2026 Elmee. All rights reserved.
//

import Foundation

internal struct DoaCategory: Codable, Hashable, Sendable {
    internal let id: String?
    internal let name: String?
    internal let slug: String?
    internal let photo: String?
}

internal struct DoaItem: Codable, Hashable, Sendable, Identifiable {
    internal var id: String {
        return _id ?? UUID().uuidString
    }
    private let _id: String?
    internal let title: String?
    internal let arabic: String?
    internal let latin: String?
    internal let translation: String?
    internal let notes: String?
    internal let fawaid: String?
    internal let source: String?
    internal let category: String?
    internal let categories: DoaCategory?

    private enum CodingKeys: String, CodingKey {
        case _id = "id"
        case title, arabic, latin, translation, notes, fawaid, source, category, categories
    }

    internal init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        _id = try container.decodeIfPresent(String.self, forKey: ._id)
        title = try container.decodeIfPresent(String.self, forKey: .title)
        arabic = try container.decodeIfPresent(String.self, forKey: .arabic)
        latin = try container.decodeIfPresent(String.self, forKey: .latin)
        notes = try container.decodeIfPresent(String.self, forKey: .notes)
        fawaid = try container.decodeIfPresent(String.self, forKey: .fawaid)
        source = try container.decodeIfPresent(String.self, forKey: .source)
        category = try container.decodeIfPresent(String.self, forKey: .category)
        categories = try container.decodeIfPresent(DoaCategory.self, forKey: .categories)

        if let str = try? container.decodeIfPresent(String.self, forKey: .translation) {
            translation = str
        } else if let dict = try? container.decodeIfPresent([String: String].self, forKey: .translation) {
            translation = dict["id"] ?? dict["ms"] ?? dict["en"]
        } else {
            translation = nil
        }
    }
}

internal struct DoaListResponse: Codable, Sendable {
    internal let success: Bool?
    internal let message: String?
    internal let data: [DoaItem]?
}

internal struct DhikrContentItem: Codable, Hashable, Sendable {
    internal let id: String?
    internal let arabic: String?
    internal let latin: String?
    internal let translation: String?
    internal let fawaid: String?
    internal let notes: String?
    internal let source: String?
    internal let repeatCount: Int?

    private enum CodingKeys: String, CodingKey {
        case repeatCount = "repeat_count"
        case id, arabic, latin, translation, fawaid, notes, source
    }

    internal init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decodeIfPresent(String.self, forKey: .id)
        arabic = try container.decodeIfPresent(String.self, forKey: .arabic)
        latin = try container.decodeIfPresent(String.self, forKey: .latin)
        fawaid = try container.decodeIfPresent(String.self, forKey: .fawaid)
        notes = try container.decodeIfPresent(String.self, forKey: .notes)
        source = try container.decodeIfPresent(String.self, forKey: .source)
        repeatCount = try container.decodeIfPresent(Int.self, forKey: .repeatCount)

        if let str = try? container.decodeIfPresent(String.self, forKey: .translation) {
            translation = str
        } else if let dict = try? container.decodeIfPresent([String: String].self, forKey: .translation) {
            translation = dict["id"] ?? dict["ms"] ?? dict["en"]
        } else {
            translation = nil
        }
    }
}

internal struct DhikrBundle: Codable, Hashable, Sendable, Identifiable {
    internal var id: String {
        return title ?? UUID().uuidString
    }
    internal let title: String?
    internal let category: String?
    internal let compiledBy: String?
    internal let content: [DhikrContentItem]?

    private enum CodingKeys: String, CodingKey {
        case compiledBy = "compiled_by"
        case title, category, content
    }
}

internal struct DhikrListResponse: Codable, Sendable {
    internal let success: Bool?
    internal let message: String?
    internal let data: [DhikrBundle]?
}

internal struct DoaCategoriesResponse: Codable, Sendable {
    internal let success: Bool?
    internal let message: String?
    internal let data: [DoaCategory]?
}

internal enum DoaCatalogKind: String, Codable, Sendable {
    case doa = "DOA"
    case dhikr = "DHIKR"
}

internal struct DoaCatalogEntry: Codable, Hashable, Sendable, Identifiable {
    internal var id: String { slug }
    internal let slug: String
    internal let title: String
    internal let kind: DoaCatalogKind
}
